controller/pods.go

```diff
@@ -51,6 +53,113 @@ const (
 	retryDelay = 1 * time.Second
 )
 
+type OpType int
+
+const (
+    Add OpType = iota
+    Delete
+)
+
+// Event struct
+type Event struct {
+	opType OpType
+	cniParams *cni.Parameters
+}
+
+// eventQueue to publish events to
+type EventQueue struct {
+	q *list.List
+	m map[string]*list.Element  // ref in the eventQueue list
+	lock sync.Mutex
+	cond *sync.Cond
+}
+
+func newQueue() *EventQueue {
+	eq := &EventQueue{m: make(map[string]*list.Element), q: list.New()}
+	eq.q.Init()
+	eq.cond = sync.NewCond(&eq.lock) 
+	return eq
+}
+
+func (eq *EventQueue) Enqueue(opType OpType, cniParams *cni.Parameters) {
+	eq.cond.L.Lock()
+	defer eq.cond.L.Unlock()
+	
+	key := eq.getKey(cniParams)
+	glog.V(5).Infof("Enqeuing using key:%s", key)
+	if e, ok := eq.m[key]; ok {
+		ev := e.Value.(Event)
+		if ev.opType != opType {
+			glog.Infof("Cancelling events from queue - events cancels each others:", ev)
+			eq.q.Remove(e)
+			delete(eq.m, key)
+			return
+		}
+    }
+	
+	ev := Event{opType, cniParams}
+	e := eq.q.PushBack(ev)
+	glog.V(5).Infof("Enqueue new event:", ev)
+	eq.m[key] = e
+	eq.cond.Signal()
+}
+
+func (eq *EventQueue) Dequeue() *Event {
+	e := eq.q.Front() // First element
+	if e == nil {
+		return nil
+	}
+	ev := e.Value.(Event)
+	eq.q.Remove(e)
+	delete(eq.m, eq.getKey(ev.cniParams))
+	return &ev
+}
+
+func (eq *EventQueue) getKey(cniParams *cni.Parameters) string {
+ 	return fmt.Sprintf("%+v", cniParams)
+}
+
+func (c *Controller) Process(e *Event) {
+	var err error
+	switch e.opType {
+	case Add:
+		for i := 0; i < maxRetries; i++ {
```

Comment:
Who said that after `maxRetries * retryDelay` seconds, (5 seconds right now) thing will works? e.g. apiserver is down.
The expectation is that if a request is done by a user to add a network attachement (via an annotation) the plumbing would eventually be done unless the user cancel that by removing the network attachement from the annotation.

Reply:
Some code was lost in commit transition.... the wait time is currently 10 minutes (100 * 10 seconds). Will update the commit.

Reply:
We could also retry forever....

Reply:
a single thread that retries forever will block all pending requests forever if the problem persists.  We need to find a way to stop retrying in situations where we know problems will never be solved (e.g. pod has died) and be able to discriminate that from situation where retrying makes sense (communication error, resources not yet ready, ...).
We should also find a way to let independent events (for example different pods) to go through if one fails.

Reply:
yes, the thread need to handle thing properly (e.g. try a couple of times on an element and if it didn't work push it back in the queue and try another element) and not hold the lock to consume the queue as it is like now.
we need also to subscribe/watch for pod delete events from apiserver to remove an element from the queue in case the pod get deleted while we are retrying.
Note:
-  a pod delete events from apiserver will be handled by kactus by a direct call from kubelet -> crio when a pod get deleted so no action w.r.t to network needs to be taken by the podagent (beside canceling an element in the queue if there is one)
- if a pod died, it would not affect the pod network (i.e. it would be restarted without destroying/creating the pod network)
-  pod re-rescheduling is different (node down, the scheduler re-start the pod on a different node) it's an add pod event that is handled by kactus directly (kubelet -> crio -> kactus) and the podagent is not involved with, if the node come back the kubelet will delete the pod on the old node and call kactus directly to clean the pod network again the podagent is not involved beside removing an element in the queue if there is one.

Reply:
In case a pod dies and is recreated on the same node, The podagent received an update event to create the network. Maybe because of BNS being involved? Same when the pod moves to another node...